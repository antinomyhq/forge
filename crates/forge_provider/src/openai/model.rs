use forge_app::domain::ModelId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Model {
    pub id: ModelId,
    pub name: Option<String>,
    pub created: Option<u64>,
    pub description: Option<String>,
    pub context_length: Option<u64>,
    pub architecture: Option<Architecture>,
    pub pricing: Option<Pricing>,
    pub top_provider: Option<TopProvider>,
    pub per_request_limits: Option<serde_json::Value>,
    pub supported_parameters: Option<Vec<String>>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Architecture {
    pub modality: String,
    pub tokenizer: String,
    pub instruct_type: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Pricing {
    pub prompt: Option<String>,
    pub completion: Option<String>,
    pub image: Option<String>,
    pub request: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct TopProvider {
    pub context_length: Option<u64>,
    pub max_completion_tokens: Option<u64>,
    pub is_moderated: bool,
}

#[derive(Debug, Deserialize, Clone, Serialize)]
pub struct ListModelResponse {
    pub data: Vec<Model>,
}
