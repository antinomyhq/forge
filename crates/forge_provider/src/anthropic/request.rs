use derive_setters::Setters;
use forge_app::domain::{ContextMessage, Image};
use serde::{Deserialize, Serialize};

use crate::error::Error;

#[derive(Serialize, Default, Setters)]
#[setters(into, strip_option)]
pub struct Request {
    max_tokens: u64,
    messages: Vec<Message>,
    model: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    metadata: Option<Metadata>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stop_sequence: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    stream: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    system: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    temperature: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tool_choice: Option<ToolChoice>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    tools: Vec<ToolDefinition>,
    #[serde(skip_serializing_if = "Option::is_none")]
    top_k: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    top_p: Option<f32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    thinking: Option<Thinking>,
}

#[derive(Serialize, Default)]
pub struct Thinking {
    r#type: String,
    budget_tokens: u64,
}

impl TryFrom<forge_app::domain::Context> for Request {
    type Error = anyhow::Error;
    fn try_from(request: forge_app::domain::Context) -> std::result::Result<Self, Self::Error> {
        // note: Anthropic only supports 1 system message in context, so from the
        // context we pick the first system message available.
        // ref: https://docs.anthropic.com/en/api/messages#body-system
        let system = request.messages.iter().find_map(|message| {
            if let ContextMessage::Text(chat_message) = message {
                if chat_message.role == forge_app::domain::Role::System {
                    Some(chat_message.content.clone())
                } else {
                    None
                }
            } else {
                None
            }
        });

        Ok(Self {
            messages: request
                .messages
                .into_iter()
                .filter(|message| {
                    // note: Anthropic does not support system messages in message field.
                    if let ContextMessage::Text(chat_message) = message {
                        chat_message.role != forge_app::domain::Role::System
                    } else {
                        true
                    }
                })
                .map(Message::try_from)
                .collect::<std::result::Result<Vec<_>, _>>()?,
            tools: request
                .tools
                .into_iter()
                .map(ToolDefinition::try_from)
                .collect::<std::result::Result<Vec<_>, _>>()?,
            system,
            temperature: request.temperature.map(|t| t.value()),
            top_p: request.top_p.map(|t| t.value()),
            top_k: request.top_k.map(|t| t.value() as u64),
            tool_choice: request.tool_choice.map(ToolChoice::from),
            thinking: request.reasoning.and_then(|reasoning| {
                match (reasoning.enabled, reasoning.max_tokens) {
                    (Some(true), Some(max_tokens)) => Some(Thinking {
                        r#type: "enabled".to_string(),
                        budget_tokens: max_tokens as u64,
                    }),
                    _ => None,
                }
            }),
            ..Default::default()
        })
    }
}

#[derive(Serialize)]
pub struct Metadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    user_id: Option<String>,
}

#[derive(Serialize)]
pub struct Message {
    content: Vec<Content>,
    role: Role,
}

impl TryFrom<ContextMessage> for Message {
    type Error = anyhow::Error;
    fn try_from(value: ContextMessage) -> std::result::Result<Self, Self::Error> {
        Ok(match value {
            ContextMessage::Text(chat_message) => {
                let mut content = Vec::with_capacity(
                    chat_message
                        .tool_calls
                        .as_ref()
                        .map(|tc| tc.len())
                        .unwrap_or_default()
                        + 1,
                );

                if let Some(reasoning) = chat_message.reasoning_details
                    && let Some((sig, text)) = reasoning.into_iter().find_map(|reasoning| {
                        match (reasoning.signature, reasoning.text) {
                            (Some(sig), Some(text)) => Some((sig, text)),
                            _ => None,
                        }
                    })
                {
                    content.push(Content::Thinking { signature: Some(sig), thinking: Some(text) });
                }

                if !chat_message.content.is_empty() {
                    // note: Anthropic does not allow empty text content.
                    content.push(Content::Text { text: chat_message.content, cache_control: None });
                }
                if let Some(tool_calls) = chat_message.tool_calls {
                    for tool_call in tool_calls {
                        content.push(tool_call.try_into()?);
                    }
                }

                match chat_message.role {
                    forge_app::domain::Role::User => Message { role: Role::User, content },
                    forge_app::domain::Role::Assistant => {
                        Message { role: Role::Assistant, content }
                    }
                    forge_app::domain::Role::System => {
                        // note: Anthropic doesn't support system role messages and they're already
                        // filtered out. so this state is unreachable.
                        return Err(Error::UnsupportedRole("System".to_string()).into());
                    }
                }
            }
            ContextMessage::Tool(tool_result) => {
                Message { role: Role::User, content: vec![tool_result.try_into()?] }
            }
            ContextMessage::Image(img) => {
                Message { content: vec![Content::from(img)], role: Role::User }
            }
        })
    }
}

impl From<Image> for Content {
    fn from(value: Image) -> Self {
        Content::Image {
            source: ImageSource {
                type_: "url".to_string(),
                media_type: None,
                data: None,
                url: Some(value.url().clone()),
            },
        }
    }
}

#[derive(Serialize)]
struct ImageSource {
    #[serde(rename = "type")]
    type_: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    media_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    data: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    url: Option<String>,
}

#[derive(Serialize)]
#[serde(rename_all = "snake_case", tag = "type")]
enum Content {
    Image {
        source: ImageSource,
    },
    Text {
        text: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        cache_control: Option<CacheControl>,
    },
    ToolUse {
        id: String,
        input: Option<serde_json::Value>,
        name: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        cache_control: Option<CacheControl>,
    },
    ToolResult {
        tool_use_id: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        content: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        is_error: Option<bool>,
        #[serde(skip_serializing_if = "Option::is_none")]
        cache_control: Option<CacheControl>,
    },
    Thinking {
        #[serde(skip_serializing_if = "Option::is_none")]
        signature: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        thinking: Option<String>,
    },
}

impl TryFrom<forge_app::domain::ToolCallFull> for Content {
    type Error = anyhow::Error;
    fn try_from(value: forge_app::domain::ToolCallFull) -> std::result::Result<Self, Self::Error> {
        let call_id = value.call_id.as_ref().ok_or(Error::ToolCallMissingId)?;

        Ok(Content::ToolUse {
            id: call_id.as_str().to_string(),
            input: serde_json::to_value(value.arguments).ok(),
            name: value.name.to_string(),
            cache_control: None,
        })
    }
}

impl TryFrom<forge_app::domain::ToolResult> for Content {
    type Error = anyhow::Error;
    fn try_from(value: forge_app::domain::ToolResult) -> std::result::Result<Self, Self::Error> {
        let call_id = value.call_id.as_ref().ok_or(Error::ToolCallMissingId)?;
        Ok(Content::ToolResult {
            tool_use_id: call_id.as_str().to_string(),
            cache_control: None,
            content: value
                .output
                .values
                .iter()
                .filter_map(|item| item.as_str().map(|s| s.to_string()))
                .next(),
            is_error: Some(value.is_error()),
        })
    }
}

#[derive(Serialize)]
#[serde(rename_all = "snake_case")]
#[allow(dead_code)]
pub enum CacheControl {
    Ephemeral,
}

#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    User,
    Assistant,
}

#[derive(Serialize)]
#[serde(rename_all = "snake_case", tag = "type")]
pub enum ToolChoice {
    Auto {
        #[serde(skip_serializing_if = "Option::is_none")]
        disable_parallel_tool_use: Option<bool>,
    },
    Any {
        #[serde(skip_serializing_if = "Option::is_none")]
        disable_parallel_tool_use: Option<bool>,
    },
    Tool {
        name: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        disable_parallel_tool_use: Option<bool>,
    },
}

// To understand the mappings refer: https://docs.anthropic.com/en/docs/build-with-claude/tool-use#controlling-claudes-output
impl From<forge_app::domain::ToolChoice> for ToolChoice {
    fn from(value: forge_app::domain::ToolChoice) -> Self {
        match value {
            forge_app::domain::ToolChoice::Auto => {
                ToolChoice::Auto { disable_parallel_tool_use: None }
            }
            forge_app::domain::ToolChoice::Call(tool_name) => {
                ToolChoice::Tool { name: tool_name.to_string(), disable_parallel_tool_use: None }
            }
            forge_app::domain::ToolChoice::Required => {
                ToolChoice::Any { disable_parallel_tool_use: None }
            }
            forge_app::domain::ToolChoice::None => {
                ToolChoice::Auto { disable_parallel_tool_use: None }
            }
        }
    }
}

#[derive(Serialize)]
pub struct ToolDefinition {
    name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    cache_control: Option<CacheControl>,
    input_schema: serde_json::Value,
}

impl TryFrom<forge_app::domain::ToolDefinition> for ToolDefinition {
    type Error = anyhow::Error;
    fn try_from(
        value: forge_app::domain::ToolDefinition,
    ) -> std::result::Result<Self, Self::Error> {
        Ok(ToolDefinition {
            name: value.name.to_string(),
            description: Some(value.description),
            cache_control: None,
            input_schema: serde_json::to_value(value.input_schema)?,
        })
    }
}
