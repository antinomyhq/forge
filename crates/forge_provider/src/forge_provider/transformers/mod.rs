mod drop_tool_call;
mod make_openai_compat;
mod pipeline;
mod set_cache;
mod tool_choice;
mod when_model;

// Use the Transformer trait from forge_domain
pub use forge_app::domain::Transformer;
pub use pipeline::ProviderPipeline;
