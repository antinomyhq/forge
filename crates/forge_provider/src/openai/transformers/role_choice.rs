use forge_app::domain::Transformer;

use crate::openai::request::{Request, Role};

/// O1 Models do not support System Roles. Only User || Assistant
pub struct RoleChoice;

impl Transformer for RoleChoice {
    type Value = Request;

    /// Since the models do not support System Role. We convert the Role to User
    /// instead
    fn transform(&mut self, mut request: Self::Value) -> Self::Value {
        if let Some(messages) = request.messages.as_mut() {
            for message in messages.iter_mut() {
                if message.role == Role::System {
                    message.role = Role::User;
                }
            }

            request
        } else {
            request
        }
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;
    use crate::openai::request::{Message, MessageContent};

    fn create_message_fixture(role: Role, content: &str) -> Message {
        Message {
            role,
            content: Some(MessageContent::Text(content.to_string())),
            name: None,
            tool_call_id: None,
            tool_calls: None,
            reasoning_details: None,
        }
    }

    fn create_request_fixture(messages: Vec<Message>) -> Request {
        Request { messages: Some(messages), ..Default::default() }
    }

    fn assert_messages_roles_eq(actual_messages: &Option<Vec<Message>>, expected_roles: &[Role]) {
        match actual_messages {
            Some(messages) => {
                assert_eq!(messages.len(), expected_roles.len());
                for (message, expected_role) in messages.iter().zip(expected_roles.iter()) {
                    assert_eq!(message.role, *expected_role);
                }
            }
            None => assert_eq!(expected_roles.len(), 0),
        }
    }

    #[test]
    fn test_transform_system_role_to_user() {
        let fixture = create_request_fixture(vec![create_message_fixture(
            Role::System,
            "You are a helpful assistant",
        )]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert_messages_roles_eq(&actual.messages, &[Role::User]);
    }

    #[test]
    fn test_preserve_user_role() {
        let fixture = create_request_fixture(vec![create_message_fixture(
            Role::User,
            "Hello, how are you?",
        )]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert_messages_roles_eq(&actual.messages, &[Role::User]);
    }

    #[test]
    fn test_preserve_assistant_role() {
        let fixture = create_request_fixture(vec![create_message_fixture(
            Role::Assistant,
            "I'm doing well, thank you!",
        )]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert_messages_roles_eq(&actual.messages, &[Role::Assistant]);
    }

    #[test]
    fn test_preserve_tool_role() {
        let fixture =
            create_request_fixture(vec![create_message_fixture(Role::Tool, "Tool response")]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert_messages_roles_eq(&actual.messages, &[Role::Tool]);
    }

    #[test]
    fn test_transform_mixed_messages() {
        let fixture = create_request_fixture(vec![
            create_message_fixture(Role::System, "You are a helpful assistant"),
            create_message_fixture(Role::User, "Hello"),
            create_message_fixture(Role::Assistant, "Hi there!"),
            create_message_fixture(Role::System, "Be concise"),
            create_message_fixture(Role::Tool, "Tool result"),
        ]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert_messages_roles_eq(
            &actual.messages,
            &[
                Role::User,      // System -> User
                Role::User,      // User stays User
                Role::Assistant, // Assistant stays Assistant
                Role::User,      // System -> User
                Role::Tool,      // Tool stays Tool
            ],
        );
    }

    #[test]
    fn test_transform_empty_messages() {
        let fixture = create_request_fixture(vec![]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert_messages_roles_eq(&actual.messages, &[]);
    }

    #[test]
    fn test_transform_none_messages() {
        let fixture = Request { messages: None, ..Default::default() };
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert!(actual.messages.is_none());
    }

    #[test]
    fn test_transform_only_system_messages() {
        let fixture = create_request_fixture(vec![
            create_message_fixture(Role::System, "First system message"),
            create_message_fixture(Role::System, "Second system message"),
        ]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        assert_messages_roles_eq(&actual.messages, &[Role::User, Role::User]);
    }

    #[test]
    fn test_transform_preserves_message_content() {
        let fixture = create_request_fixture(vec![create_message_fixture(
            Role::System,
            "You are a helpful assistant",
        )]);
        let mut transformer = RoleChoice;

        let actual = transformer.transform(fixture);

        let messages = actual.messages.unwrap();
        assert_eq!(messages.len(), 1);
        assert_eq!(messages[0].role, Role::User);
        if let Some(MessageContent::Text(content)) = &messages[0].content {
            assert_eq!(content, "You are a helpful assistant");
        } else {
            panic!("Expected text content");
        }
    }
}
