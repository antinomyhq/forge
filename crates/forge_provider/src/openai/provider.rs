use anyhow::{Context as _, Result};
use derive_builder::Builder;
use forge_app::domain::{
    ChatCompletionMessage, Context as ChatContext, ModelId, Provider, ContextMessage, ResultStream, Content
};
use reqwest::header::{AUTHORIZATION, HeaderMap, HeaderValue};
use reqwest::{Client, Url};
use reqwest_eventsource::{Event, RequestBuilderExt};
use tokio_stream::StreamExt;
use tracing::{debug, info};
use serde_json::{json, Value};
use super::model::{ListModelResponse, Model};
use super::request::Request;
use super::response::Response;
use crate::error::Error;
use crate::openai::transformers::{ProviderPipeline, Transformer};
use crate::utils::{format_http_context, sanitize_headers};
use std::sync::Mutex;
use once_cell::sync::Lazy;
use std::collections::HashMap;

#[derive(Clone, Builder)]
pub struct ForgeProvider {
    client: Client,
    provider: Provider,
    version: String,
}

static THREAD_ID_CACHE: Lazy<Mutex<HashMap<String, String>>> = Lazy::new(|| Mutex::new(HashMap::new()));

impl ForgeProvider {
    pub fn builder() -> ForgeProviderBuilder {
        ForgeProviderBuilder::default()
    }

    fn url(&self, path: &str) -> anyhow::Result<Url> {
        // Validate the path doesn't contain certain patterns
        if path.contains("://") || path.contains("..") {
            anyhow::bail!("Invalid path: Contains forbidden patterns");
        }

        // Remove leading slash to avoid double slashes
        let path = path.trim_start_matches('/');

        self.provider.to_base_url().join(path).with_context(|| {
            format!(
                "Failed to append {} to base URL: {}",
                path,
                self.provider.to_base_url()
            )
        })
    }

    async fn copilot_create_thread(&self) -> anyhow::Result<String> {
        let url = self.url("github/chat/threads")?;
        let headers = self.headers();
        let resp = self.client.post(url)
            .headers(headers)
            .send()
            .await?
            .error_for_status()?
            .json::<serde_json::Value>()
            .await?;
        // Extract thread_id from response
        let thread_id = resp.get("thread_id")
            .and_then(|v| v.as_str())
            .ok_or_else(|| anyhow::anyhow!("No thread id in Copilot response"))?;
        Ok(thread_id.to_string())
    }

    // OpenRouter optional headers ref: https://openrouter.ai/docs/api-reference/overview#headers
    // - `HTTP-Referer`: Identifies your app on openrouter.ai
    // - `X-Title`: Sets/modifies your app's title
    fn headers(&self) -> HeaderMap {
        let mut headers = HeaderMap::new();
        if let Some(ref api_key) = self.provider.key() {
            if self.provider.is_copilot() {
                headers.insert(
                    AUTHORIZATION,
                    HeaderValue::from_str(&format!("GitHub-Bearer {api_key}")).unwrap(),
                );            
                headers.insert(
                    "Copilot-Integration-Id",
                    HeaderValue::from_static("forge"),
                );
            } else {
                headers.insert(
                    AUTHORIZATION,
                    HeaderValue::from_str(&format!("Bearer {api_key}")).unwrap(),
                );
            }
        }

        headers.insert("X-Title", HeaderValue::from_static("forge"));
        headers.insert(
            "x-app-version",
            HeaderValue::from_str(format!("v{}", self.version).as_str())
                .unwrap_or(HeaderValue::from_static("v0.1.0-dev")),
        );
        headers.insert(
            "HTTP-Referer",
            HeaderValue::from_static("https://github.com/antinomyhq/forge"),
        );
        headers.insert(
            reqwest::header::CONNECTION,
            HeaderValue::from_static("keep-alive"),
        );
        debug!(headers = ?sanitize_headers(&headers), "Request Headers");
        headers
    }

    

    async fn copilot_chat(
        &self,
        model: &ModelId,
        context: ChatContext,
        thread_id: String,
    ) -> ResultStream<ChatCompletionMessage, anyhow::Error> {
        // 1. Extract the latest user message content
        let user_content = context
            .messages
            .last()
            .map(|m| m.to_text().clone())
            .unwrap_or_else(|| "".to_string());
    
        // 2. Build the Copilot request payload
        let payload = json!({
            "content": user_content,
            "context": context.messages.iter().map(|m| {
                match m {
                    ContextMessage::Text(text_message) => {
                        json!({
                            "role": text_message.role.to_string(),
                            "content": text_message.content,
                        })
                    },
                    ContextMessage::Tool(tool_message) => {
                        json!({
                            "role": "tool",
                            "content": format!("{:?}", tool_message.output),
                        })
                    },
                    ContextMessage::Image(image_message) => {
                        json!({
                            "role": "user",
                            "content": "[image omitted]",
                        })
                    }
                }
            }).collect::<Vec<_>>(),
            "intent": "conversation",
            "model": model.as_str(),
            "thread_id": thread_id,
            "streaming": true,
        });
    
        // 3. Build the URL
        let url = self.url(&format!("github/chat/threads/{}/messages", thread_id))?;
        let headers = self.headers();
        let url_for_filter = url.clone();
    
        // 4. Make the POST request and get the event stream
        let es = self
            .client
            .post(url.clone())
            .headers(headers)
            .json(&payload)
            .eventsource()
            .with_context(|| format_http_context(None, "POST", &url))?;
    
        // 5. Parse the event stream
        let stream = es
            .take_while(|message| {
                let is_stream_ended = matches!(message, Err(reqwest_eventsource::Error::StreamEnded));
                !is_stream_ended
            })
            .then(move |event| {
                async move {
                    match event {
                        Ok(Event::Message(ev)) => {
                            if let Ok(json) = serde_json::from_str::<Value>(&ev.data) {
                                match json.get("type").and_then(|t| t.as_str()) {
                                    Some("content") => {
                                        if let Some(body) = json.get("body").and_then(|b| b.as_str()) {
                                            let content = Content::part(body);
                                            Some(Ok(ChatCompletionMessage::assistant(content.clone())))
                                        } else {
                                            None
                                        }
                                    }
                                    Some("complete") => {
                                        // Stream is complete, no more messages to send
                                        None
                                    }
                                    _ => None,
                                }
                            } else {
                                None
                            }
                        }
                        Ok(Event::Open) => {
                            None
                        }
                        Err(error) => {
                            match error {
                                reqwest_eventsource::Error::StreamEnded => None,
                                reqwest_eventsource::Error::InvalidStatusCode(_, response) => {
                                    let status = response.status();
                                    Some(Err(anyhow::anyhow!(Error::InvalidStatusCode(status.as_u16()))
                                        .context(format!("HTTP {}", status))))
                                }
                                reqwest_eventsource::Error::InvalidContentType(_, ref response) => {
                                    let status_code = response.status();
                                    Some(Err(anyhow::anyhow!(error)
                                        .context(format!("Invalid content type. HTTP Status: {}", status_code))))
                                }
                                error => {
                                    tracing::error!(error = ?error, "Failed to receive chat completion event");
                                    Some(Err(anyhow::anyhow!(error)))
                                }
                            }
                        }
                    }
                }
            })
            .filter_map(move |response| {
                response.map(|result| result.with_context(|| format_http_context(None, "POST", &url_for_filter)))
            });
        Ok(Box::pin(stream))
    }   

    async fn inner_chat(
        &self,
        model: &ModelId,
        context: ChatContext,
    ) -> ResultStream<ChatCompletionMessage, anyhow::Error> 
    {
        if self.provider.is_copilot() {
            let cid = context.conversation_id.clone().unwrap_or_default().to_string();
            // Check cache for thread_id
            let cached_tid = {
                let cache = THREAD_ID_CACHE.lock().unwrap();
                cache.get(&cid).cloned()
            };
            let thread_id = if let Some(tid) = cached_tid {
                tid
            } else {
                // Create thread outside lock
                let new_tid = self.copilot_create_thread().await?;
                // Insert into cache
                let mut cache = THREAD_ID_CACHE.lock().unwrap();
                cache.insert(cid, new_tid.clone());
                new_tid
            };
            return self.copilot_chat(model, context.clone(), thread_id).await;
        }
        let mut request = Request::from(context.clone()).model(model.clone()).stream(true);
        let mut pipeline = ProviderPipeline::new(&self.provider);
        request = pipeline.transform(request);
    
        let url = self.url("chat/completions")?;
        let headers = self.headers();

        info!(
            url = %url,
            model = %model,
            headers = ?sanitize_headers(&headers),
            message_count = %request.message_count(),
            message_cache_count = %request.message_cache_count(),
            "Connecting Upstream"
        );

        let es = self
            .client
            .post(url.clone())
            .headers(headers)
            .json(&request)
            .eventsource()
            .with_context(|| format_http_context(None, "POST", &url))?;
        let stream = es
            .take_while(|message| !matches!(message, Err(reqwest_eventsource::Error::StreamEnded)))
            .then(|event| async {
                match event {
                    Ok(event) => match event {
                        Event::Open => None,
                        Event::Message(event) if ["[DONE]", ""].contains(&event.data.as_str()) => {
                            debug!("Received completion from Upstream");
                            None
                        }
                        Event::Message(message) => Some(
                            serde_json::from_str::<Response>(&message.data)
                                .with_context(|| {
                                    format!(
                                        "Failed to parse Forge Provider response: {}",
                                        message.data
                                    )
                                })
                                .and_then(|response| {
                                    ChatCompletionMessage::try_from(response.clone()).with_context(
                                        || {
                                            format!(
                                                "Failed to create completion message: {}",
                                                message.data
                                            )
                                        },
                                    )
                                }),
                        ),
                    },
                    Err(error) => match error {
                        reqwest_eventsource::Error::StreamEnded => None,
                        reqwest_eventsource::Error::InvalidStatusCode(_, response) => {
                            let status = response.status();
                            let body = response.text().await.ok();
                            Some(Err(Error::InvalidStatusCode(status.as_u16())).with_context(
                                || match body {
                                    Some(body) => {
                                        format!("{status} Reason: {body}")
                                    }
                                    None => {
                                        format!("{status} Reason: [Unknown]")
                                    }
                                },
                            ))
                        }
                        reqwest_eventsource::Error::InvalidContentType(_, ref response) => {
                            let status_code = response.status();
                            debug!(response = ?response, "Invalid content type");
                            Some(Err(error).with_context(|| format!("Http Status: {status_code}")))
                        }
                        error => {
                            tracing::error!(error = ?error, "Failed to receive chat completion event");
                            Some(Err(error.into()))
                        }
                    },
                }
            })
            .filter_map(move |response| {
                response
                    .map(|result| result.with_context(|| format_http_context(None, "POST", &url)))
            });

        Ok(Box::pin(stream))
    }

    async fn inner_models(&self) -> Result<Vec<forge_app::domain::Model>> {
        let url = self.url("models")?;
        debug!(url = %url, "Fetching models");
        match self.fetch_models(url.clone()).await {
            Err(error) => {
                tracing::error!(error = ?error, "Failed to fetch models");
                anyhow::bail!(error)
            }
            Ok(response) => {
                let data: ListModelResponse = serde_json::from_str(&response)
                    .with_context(|| format_http_context(None, "GET", &url))
                    .with_context(|| "Failed to deserialize models response")?;
                Ok(data.data.into_iter().map(Into::into).collect())
            }
        }
    }

    async fn fetch_models(&self, url: Url) -> Result<String, anyhow::Error> {
        let headers = self.headers();
        info!(method = "GET", url = %url, headers = ?sanitize_headers(&headers), "Fetching Models");
        match self.client.get(url.clone()).headers(headers).send().await {
            Ok(response) => {
                let status = response.status();
                let ctx_message = format_http_context(Some(status), "GET", &url);
                let response = response
                    .text()
                    .await
                    .with_context(|| ctx_message.clone())
                    .with_context(|| "Failed to decode response into text")?;
                if status.is_success() {
                    Ok(response)
                } else {
                    // treat non 200 response as error.
                    Err(anyhow::anyhow!(response))
                        .with_context(|| ctx_message)
                        .with_context(|| "Failed to fetch the models")
                }
            }
            Err(err) => {
                let ctx_msg = format_http_context(err.status(), "GET", &url);
                Err(err)
                    .with_context(|| ctx_msg)
                    .with_context(|| "Failed to fetch the models")
            }
        }
    }
}

impl ForgeProvider {
    pub async fn chat(
        &self,
        model: &ModelId,
        context: ChatContext,
    ) -> ResultStream<ChatCompletionMessage, anyhow::Error> {
        self.inner_chat(model, context).await
    }

    pub async fn models(&self) -> Result<Vec<forge_app::domain::Model>> {
        self.inner_models().await
    }
}

impl From<Model> for forge_app::domain::Model {
    fn from(value: Model) -> Self {
        let tools_supported = value
            .supported_parameters
            .iter()
            .flatten()
            .any(|param| param == "tools");
        let supports_parallel_tool_calls = value
            .supported_parameters
            .iter()
            .flatten()
            .any(|param| param == "supports_parallel_tool_calls");
        let is_reasoning_supported = value
            .supported_parameters
            .iter()
            .flatten()
            .any(|param| param == "reasoning");

        forge_app::domain::Model {
            id: value.id,
            name: value.name,
            description: value.description,
            context_length: value.context_length,
            tools_supported: Some(tools_supported),
            supports_parallel_tool_calls: Some(supports_parallel_tool_calls),
            supports_reasoning: Some(is_reasoning_supported),
        }
    }
}

#[cfg(test)]
mod tests {
    use anyhow::Context;
    use reqwest::Client;

    use super::*;
    use crate::mock_server::{MockServer, normalize_ports};

    fn create_provider(base_url: &str) -> anyhow::Result<ForgeProvider> {
        let provider = Provider::OpenAI {
            url: reqwest::Url::parse(base_url)?,
            key: Some("test-api-key".to_string()),
        };

        Ok(ForgeProvider::builder()
            .client(Client::new())
            .provider(provider)
            .version("1.0.0".to_string())
            .build()
            .unwrap())
    }

    fn create_mock_models_response() -> serde_json::Value {
        serde_json::json!({
            "data": [
                {
                    "id": "model-1",
                    "name": "Test Model 1",
                    "description": "A test model",
                    "context_length": 4096,
                    "supported_parameters": ["tools", "supports_parallel_tool_calls"]
                },
                {
                    "id": "model-2",
                    "name": "Test Model 2",
                    "description": "Another test model",
                    "context_length": 8192,
                    "supported_parameters": ["tools"]
                }
            ]
        })
    }

    fn create_error_response(message: &str, code: u16) -> serde_json::Value {
        serde_json::json!({
            "error": {
                "message": message,
                "code": code
            }
        })
    }

    fn create_empty_response() -> serde_json::Value {
        serde_json::json!({ "data": [] })
    }

    #[tokio::test]
    async fn test_fetch_models_success() -> anyhow::Result<()> {
        let mut fixture = MockServer::new().await;
        let mock = fixture
            .mock_models(create_mock_models_response(), 200)
            .await;
        let provider = create_provider(&fixture.url())?;
        let actual = provider.models().await?;

        mock.assert_async().await;
        insta::assert_json_snapshot!(actual);
        Ok(())
    }

    #[tokio::test]
    async fn test_fetch_models_http_error_status() -> anyhow::Result<()> {
        let mut fixture = MockServer::new().await;
        let mock = fixture
            .mock_models(create_error_response("Invalid API key", 401), 401)
            .await;

        let provider = create_provider(&fixture.url())?;
        let actual = provider.models().await;

        mock.assert_async().await;

        // Verify that we got an error
        assert!(actual.is_err());
        insta::assert_snapshot!(normalize_ports(format!("{:#?}", actual.unwrap_err())));
        Ok(())
    }

    #[tokio::test]
    async fn test_fetch_models_server_error() -> anyhow::Result<()> {
        let mut fixture = MockServer::new().await;
        let mock = fixture
            .mock_models(create_error_response("Internal Server Error", 500), 500)
            .await;

        let provider = create_provider(&fixture.url())?;
        let actual = provider.models().await;

        mock.assert_async().await;

        // Verify that we got an error
        assert!(actual.is_err());
        insta::assert_snapshot!(normalize_ports(format!("{:#?}", actual.unwrap_err())));
        Ok(())
    }

    #[tokio::test]
    async fn test_fetch_models_empty_response() -> anyhow::Result<()> {
        let mut fixture = MockServer::new().await;
        let mock = fixture.mock_models(create_empty_response(), 200).await;

        let provider = create_provider(&fixture.url())?;
        let actual = provider.models().await?;

        mock.assert_async().await;
        assert!(actual.is_empty());
        Ok(())
    }

    #[test]
    fn test_error_deserialization() -> Result<()> {
        let content = serde_json::to_string(&serde_json::json!({
          "error": {
            "message": "This endpoint's maximum context length is 16384 tokens",
            "code": 400
          }
        }))
        .unwrap();
        let message = serde_json::from_str::<Response>(&content)
            .with_context(|| "Failed to parse response")?;
        let message = ChatCompletionMessage::try_from(message.clone());

        assert!(message.is_err());
        Ok(())
    }
}