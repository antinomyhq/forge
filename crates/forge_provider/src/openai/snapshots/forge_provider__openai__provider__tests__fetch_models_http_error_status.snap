---
source: crates/forge_provider/src/openai/provider.rs
expression: "normalize_ports(format!(\"{:#?}\", actual.unwrap_err()))"
---
Error {
    context: "Failed to fetch the models",
    source: Error {
        context: "401 GET http://127.0.0.1:<port>/models",
        source: Object {
            "error": Object {
                "code": Number(401),
                "message": String("Invalid API key"),
            },
        },
    },
}
